sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePartitionedUploader.cs

```diff
@@ -0,0 +1,370 @@
+
+                // Calculate flush position
+                long flushPosition = offset + contentLength;
+
+                return _client.Flush(
+                    position: flushPosition,
+                    httpHeaders: httpHeaders);
+            }
+
+            // If the caller provided an explicit block size, we'll use it.
+            // Otherwise we'll adjust dynamically based on the size of the
+            // content.
+            int blockSize =
+                _blockSize != null ? _blockSize.Value :
+                contentLength < Constants.LargeUploadThreshold ?
+                    Constants.DefaultBufferSize :
+                    Constants.LargeBufferSize;
+
+            // Otherwise stage individual blocks one at a time.  It's not as
+            // fast as a parallel upload, but you get the benefit of the retry
+            // policy working on a single block instead of the entire stream.
+            return UploadInSequence(
+                content,
+                blockSize,
+                offset,
+                httpHeaders,
+                conditions,
+                progressHandler,
+                cancellationToken);
+        }
+
+        private Response<PathInfo> UploadInSequence(
+            Stream content,
+            int blockSize,
+            long offset,
+            PathHttpHeaders httpHeaders,
+            DataLakeRequestConditions conditions,
+            IProgress<long> progressHandler,
+            CancellationToken cancellationToken)
+        {
+            // Wrap the append and flush calls in an Upload span for
+            // distributed tracing
+            DiagnosticScope scope = _client.ClientDiagnostics.CreateScope(
+                _operationName ?? $"{nameof(Azure)}.{nameof(Storage)}.{nameof(Files)}.{nameof(DataLake)}.{nameof(DataLakeFileClient)}.{nameof(DataLakeFileClient.Upload)}");
+
+            try
+            {
+                scope.Start();
+
+                // Wrap progressHandler in a AggregatingProgressIncrementer to prevent
+                // progress from being reset with each append file operation.
+                if (progressHandler != null)
+                {
+                    progressHandler = new AggregatingProgressIncrementer(progressHandler);
+                }
+
+                // Partition the stream into individual blocks and stage them
+                IAsyncEnumerator<ChunkedStream> enumerator =
+                    PartitionedUploadExtensions.GetBlocksAsync(
+                        content, blockSize, async: false, _arrayPool, cancellationToken)
+                    .GetAsyncEnumerator(cancellationToken);
+
+                // We need to keep track of how much data we have appended to
+                // calculate offsets for the next appends, and the final
+                // position to flush
+                long appendedBytes = 0;
+
+                while (enumerator.MoveNextAsync().EnsureCompleted())
```

Comment:
if this does not complete successfully, does this throw an exception or does it exit the while loop?

Reply:
Unsure, I'm following the logic of the Blob version of the PartitionedUploader.